lib/client.js

```diff
@@ -689,44 +689,80 @@ Client.prototype._getEncoder = function () {
  * @private
  */
 Client.prototype._setQueryOptions = function (options, params, meta, callback) {
-  var paramsInfo;
-  var self = this;
-  function adaptPageStateAndRouting(keys, params) {
-    if (typeof options.pageState === 'string') {
-      //pageState can be a hex string
-      options.pageState = new Buffer(options.pageState, 'hex');
-    }
-    self._getEncoder().setRoutingKey(params, options, keys);
-  }
-  if (!options.prepare && params && !util.isArray(params) && this.controlConnection.protocolVersion < 3) {
+  var protocolVersion = this.controlConnection.protocolVersion;
+  if (!options.prepare && params && !util.isArray(params) && protocolVersion < 3) {
     //Only Cassandra 2.1 and above supports named parameters
     return callback(new errors.ArgumentError('Named parameters for simple statements are not supported, use prepare flag'));
   }
-  try {
-    if (options.prepare) {
-      paramsInfo = utils.adaptNamedParamsPrepared(params, meta.columns);
-      //Add the type information provided by the prepared metadata
-      options.hints = meta.columns.map(function (c) { return c.type; });
-      adaptPageStateAndRouting(paramsInfo.keys, paramsInfo.params);
+  var paramsInfo;
+  var self = this;
+  async.series([
+    function fillRoutingKeys(next) {
+      if (options.routingKey || options.routingIndexes || options.routingNames || !meta) {
+        //it is filled by the user
+        //or it is not prepared
+        return next();
+      }
+      if (util.isArray(meta.partitionKeys)) {
+        //the partition keys are provided as part of the metadata
+        options.routingIndexes = meta.partitionKeys;
+        return next();
+      }
+      self.metadata.getTable(meta.keyspace, meta.table, function (err, tableInfo) {
```

Comment:
Woah this is pretty cool, is this new (from what I can tell it is)?   This would only apply to proto version < 4 right? ( since meta.partitionKeys will be present).